src/Evolution/Systems/ScalarWave/Characteristics.cpp

```diff
@@ -0,0 +1,193 @@
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        unit_normal_one_form) noexcept {
+  auto char_speeds =
+      make_with_value<typename Tags::CharacteristicSpeeds<Dim>::type>(
+          get<0>(unit_normal_one_form), 0.);
+  characteristic_speeds(make_not_null(&char_speeds), unit_normal_one_form);
+  return char_speeds;
+}
+
+template <size_t Dim>
+void characteristic_fields(
+    const gsl::not_null<typename Tags::CharacteristicFields<Dim>::type*>
+        char_fields,
+    const Scalar<DataVector>& gamma_2, const Scalar<DataVector>& psi,
+    const Scalar<DataVector>& pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& phi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        unit_normal_one_form) noexcept {
+  // n^j = \delta^{jk} n_k
+  auto unit_normal_vector =
+      make_with_value<tnsr::I<DataVector, Dim, Frame::Inertial>>(pi, 0.);
+  for (size_t i = 0; i < Dim; ++i) {
+    unit_normal_vector.get(i) = unit_normal_one_form.get(i);
+  }
+
+  // Compute phi_dot_normal = n^i \Phi_{i}
+  auto phi_dot_normal = make_with_value<Scalar<DataVector>>(pi, 0.);
+  for (size_t i = 0; i < Dim; ++i) {
+    get(phi_dot_normal) += unit_normal_vector.get(i) * phi.get(i);
+  }
+
+  // Eq.(34) of Holst+ (2004) for UZero
+  for (size_t i = 0; i < Dim; ++i) {
+    get<Tags::UZero<Dim>>(*char_fields).get(i) =
+        phi.get(i) - unit_normal_one_form.get(i) * get(phi_dot_normal);
+  }
+
+  // Eq.(33) of Holst+ (2004) for UPsi
+  get<Tags::UPsi>(*char_fields) = psi;
+
+  // Eq.(35) of Holst+ (2004) for UPlus and UMinus
+  get(get<Tags::UPlus>(*char_fields)) =
+      get(pi) + get(phi_dot_normal) - get(gamma_2) * get(psi);
+  get(get<Tags::UMinus>(*char_fields)) =
+      get(pi) - get(phi_dot_normal) - get(gamma_2) * get(psi);
+}
+
+template <size_t Dim>
+typename Tags::CharacteristicFields<Dim>::type characteristic_fields(
+    const Scalar<DataVector>& gamma_2, const Scalar<DataVector>& psi,
+    const Scalar<DataVector>& pi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& phi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        unit_normal_one_form) noexcept {
+  auto char_fields =
+      make_with_value<typename Tags::CharacteristicFields<Dim>::type>(
+          get(gamma_2), 0.);
+  characteristic_fields(make_not_null(&char_fields), gamma_2, psi, pi, phi,
+                        unit_normal_one_form);
+  return char_fields;
+}
+
+template <size_t Dim>
+void evolved_fields_from_characteristic_fields(
+    gsl::not_null<
+        typename Tags::EvolvedFieldsFromCharacteristicFields<Dim>::type*>
+        evolved_fields,
+    const Scalar<DataVector>& gamma_2, const Scalar<DataVector>& u_psi,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& u_zero,
+    const Scalar<DataVector>& u_plus, const Scalar<DataVector>& u_minus,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        unit_normal_one_form) noexcept {
+  // Eq.(36) of Holst+ (2005) for Psi
```

Comment:
Is 2005 here  a typo?

Reply:
yes, fixed, thanks. should've been 2004.